File: leetcode/1001-2000/1101-1200/1161-1170/1169-maximum-level-sum-of-a-binary-tree/MaximumLevelSumOfABinaryTree.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

function maxLevelSum(root: TreeNode | null): number {
  if (root == null) {
    return 0;
  }

  const sums: number[] = [];

  const bfs = (node: TreeNode, depth: number): void => {
    if (depth === sums.length) {
      sums.push(node.val);
    } else {
      sums[depth] += node.val;
    }

    if (node.left) {
      bfs(node.left, depth + 1);
    }

    if (node.right) {
      bfs(node.right, depth + 1)
    }
  }

  bfs(root, 0);

  let level = 0;
  let max = sums[0];

  for (let i = 1; i < sums.length; i++) {
    if (max < sums[i]) {
      max = sums[i];
      level = i;
    }
  }

  return level + 1;
};

const root = new TreeNode(
  1,
  new TreeNode(7, new TreeNode(7), new TreeNode(-8)),
  new TreeNode(0));

console.log(maxLevelSum(root));

File: leetcode/1001-2000/1101-1200/1161-1170/1169-maximum-level-sum-of-a-binary-tree/MaximumLevelSumOfABinaryTree.java
import java.util.ArrayList;
import java.util.List;

public class MaximumLevelSumOfABinaryTree {
  public static void main(String[] args) {
    TreeNode root = new TreeNode(
        1,
        new TreeNode(7, new TreeNode(7), new TreeNode(-8)),
        new TreeNode(0));

    System.out.println(new Solution().maxLevelSum(root));
  }
}

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {
  }

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  void sum(TreeNode node, int depth, List<Long> sums) {
    if (sums.size() == depth) {
      sums.add((long) node.val);
    } else {
      sums.set(depth, sums.get(depth) + node.val);
    }

    if (node.left != null) {
      sum(node.left, depth + 1, sums);
    }

    if (node.right != null) {
      sum(node.right, depth + 1, sums);
    }
  }

  public int maxLevelSum(TreeNode root) {
    if (root == null) {
      return 0;
    }

    List<Long> sums = new ArrayList<>();

    sum(root, 0, sums);

    int maxLevelSum = 0;
    long maxSum = sums.get(0);

    for (int i = 1; i < sums.size(); i++) {
      if (maxSum < sums.get(i)) {
        maxSum = sums.get(i);
        maxLevelSum = i;
      }
    }

    return maxLevelSum + 1;
  }
}
